src/core/metrics.cc

```diff
@@ -163,205 +183,290 @@ Metrics::EnableGPUMetrics()
+              if (energy_fail_cnt[didx] < fail_threshold) {
+                unsigned int energy = field_values[2].value.i64;
+                if ((field_values[2].status == DCGM_ST_OK) &&
+                    (!DCGM_INT64_IS_BLANK(energy))) {
+                  energy_fail_cnt[didx] = 0;
+                  if (last_energy[didx] == 0) {
+                    last_energy[didx] = energy;
+                  }
+                  gpu_energy_consumption_[didx]->Increment(
+                      (double)(energy - last_energy[didx]) * 0.001);
+                  last_energy[didx] = energy;
+                } else {
+                  energy_fail_cnt[didx]++;
+                  energy = 0;
+                  LOG_WARNING << "Unable to get energy consumption for "
+                              << "GPU " << cuda_id << ": "
+                              << errorString(dcgmerr);
+                }
+              }
 
-          // Memory
-          if (mem_fail_cnt[didx] < fail_threshold) {
-            nvmlMemory_t mem;
-            nvmlReturn_t nvmlerr = nvmlDeviceGetMemoryInfo(gpu, &mem);
-            if (nvmlerr == NVML_SUCCESS) {
-              mem_fail_cnt[didx] = 0;
-            } else {
-              LOG_WARNING << "failed to get memory for GPU " << didx << ": "
-                          << nvmlErrorString(nvmlerr);
-              mem.total = 0;
-              mem.used = 0;
-              mem_fail_cnt[didx]++;
+              // Utilization
+              if (util_fail_cnt[didx] < fail_threshold) {
+                unsigned int util = field_values[3].value.i64;
+                if ((field_values[3].status == DCGM_ST_OK) &&
+                    (!DCGM_INT64_IS_BLANK(util))) {
+                  util_fail_cnt[didx] = 0;
+                } else {
+                  util_fail_cnt[didx]++;
+                  util = 0;
+                  LOG_WARNING << "Unable to get GPU utilization for GPU "
+                              << cuda_id << ": " << errorString(dcgmerr);
+                }
+                gpu_utilization_[didx]->Set((double)util * 0.01);
+              }
+
+              // Memory Usage
+              if (mem_fail_cnt[didx] < fail_threshold) {
+                unsigned int memory_used = field_values[4].value.i64;
+                unsigned int memory_total = field_values[5].value.i64;
+                if ((field_values[4].status == DCGM_ST_OK) &&
+                    (!DCGM_INT64_IS_BLANK(memory_used)) &&
+                    (field_values[5].status == DCGM_ST_OK) &&
+                    (!DCGM_INT64_IS_BLANK(memory_total))) {
+                  mem_fail_cnt[didx] = 0;
+                } else {
+                  memory_total = 0;
+                  memory_used = 0;
+                  mem_fail_cnt[didx]++;
+                  LOG_WARNING << "Unable to get memory usage for GPU "
+                              << cuda_id << ": " << errorString(dcgmerr);
+                }
+                gpu_memory_total_[didx]->Set(memory_total * 1e6);  // bytes
```

Comment:
Don't use "1e6" because that is a float value which then does math as floats and converts back to int. Use "1000 * 1000". Or likely you need to use "1024 * 1024" because the field value is in MiB... you need to check that.